addon/serializers/json.js

```diff
@@ -843,14 +843,35 @@ var JSONSerializer = Serializer.extend({
   /**
     Check if the given hasMany relationship should be serialized
 
-    @method _shouldSerializeHasMany
+    @method shouldSerializeHasMany
+    @param {DS.Snapshot} snapshot
+    @param {String} key
+    @param {String} relationshipType
+    @return {boolean} true if the hasMany relationship should be serialized
+  */
+  shouldSerializeHasMany(snapshot, key, relationship) {
+    return this.__shouldSerializeHasMany(snapshot, key, relationship);
```

Comment:
In here we can check if `this._shouldSerializeHasMany === JSONSerializer.prototype._shouldSerializeHasMany` and only issue a deprecation warning and call `_shouldSerializeHasMany` if they are not equal. See https://github.com/emberjs/data/blob/master/addon/serializers/rest.js#L755-L773 for an example.